plugins/woocommerce/includes/wc-cart-functions.php

```diff
@@ -179,21 +182,33 @@ function wc_clear_cart_after_payment() {
 			$order = wc_get_order( $order_id );
 
 			if ( $order instanceof WC_Order && hash_equals( $order->get_order_key(), $order_key ) ) {
-				WC()->cart->empty_cart();
+				$should_clear_cart_after_payment = true;
 			}
 		}
 	}
 
-	if ( is_object( WC()->session ) && WC()->session->order_awaiting_payment > 0 ) {
+	// If the order is awaiting payment, and we haven't already decided to clear the cart, check the order status.
+	if ( is_object( WC()->session ) && WC()->session->order_awaiting_payment > 0 && ! $should_clear_cart_after_payment ) {
 		$order = wc_get_order( WC()->session->order_awaiting_payment );
 
 		if ( $order instanceof WC_Order && $order->get_id() > 0 ) {
-			// If the order has not failed, or is not pending, the order must have gone through.
-			if ( ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) ) ) {
-				WC()->cart->empty_cart();
-			}
+			// If the order status is neither pending, failed, nor cancelled, the order must have gone through.
+			$did_order_succeed               = ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) );
+			$should_clear_cart_after_payment = $did_order_succeed;
```

Comment:
**(3/4)** Similarly here, we believe a payment took place. Let's make the same update (I'm also collapsing the two existing lines into one, but we don't have to do that).

```suggestion
			$should_clear_cart_after_payment = ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) );
			$after_payment                   = true;
```